### PR TITLE
Alter handling of c_ptr arguments in Python interoperability

### PR DESCRIPTION
**Pull request by lydia-duncan**
Prior to this change, c_ptr arguments would be left mostly unaltered for the
generated Python files.  This was a problem, because Python doesn't have a
notion of a pointer type - when trying to compile the generated Python files, we
would encounter failures because the type wasn't recognizable to Python.

Here, I've taken the approach that a pointer argument for an exported function
can be treated like a numpy array argument containing that type (in which case
we pass along a pointer to the internal storage for the numpy array) or a ctype
pointer.  This does not explicitly check against their types, but I think that would
be reasonable future work.  ctypes.byref cannot be supported so far as I can
tell, and other types lead to an error message now.

This strategy is needed to support returning c_strings whose memory has been
allocated, as we cannot easily detect this from the code (and so leak memory if
it is returned directly - the return value is copied into a Python bytes string
and the value handled within the generated .pyx, so without the ability to
detect if the memory must be explicitly freed the reference will go out of scope
without having been cleaned up).  Now, the string contents can be copied into
memory provided to the function via the pointer argument, without making the
direct return of c_strings more difficult.

Add a test locking in the pointer behavior, demonstrating the string arg return using
numpy arrays.  Also adds four tests of the ctypes pointer behavior, to ensure that
works as well.  The string arg return case mocks up the intended strategy for
returning c_strings that came from Chapel strings (as opposed to c_strings that
were always c_strings).

In summary: the test creates an array in Python to store the information you want
to get back as individual characters.  Pass a pointer to this array and the size
allocated for it to the function.  In the function, memcpy the contents of the
string into this array and return the size used.  Once the function has
returned, remove the excess space from the array, and convert the characters
from their character code back into a string.

Passed testing of the python interop directory hierarchy with futures

### CODE REVIEW

**Comment by lydia-duncan** [comment]
I'm trying to make an example using the ctypes pointers.  So far it is being finicky, which is obnoxious

File: compiler/codegen/symbol.cpp [hunk 1]
```diff
@@ -1090,6 +1108,53 @@ std::string ArgSymbol::getPythonArgTranslation() {
 
       return res;
     }
+  } else if (t->symbol->hasEitherFlag(FLAG_C_PTR_CLASS, FLAG_REF)) {
```
Comment (mppf) [inline compiler/codegen/symbol.cpp]: Allowing one to pass a numpy array into an export proc expecting a ref seems strange but I don't think it's inherently wrong

**Comment by lydia-duncan** [comment]
I'd kinda like to add a template string to use for the generated python output, where I have a symbol (%arg or something) that can have all instances replaced with the relevant argument name instead of writing all of it out by hand in a format that is not very understandable to the developer.  I think I'm a little too close to the AI workflow code freeze for that, so I'd like to merge this first and do that
change if I have time.

**Comment by lydia-duncan** [comment]
I believe when I've done similar things in the past, it was all in Python, but I think I do something similar in C++

**Comment by lydia-duncan** [comment]
Anyways, I think this is ready for a second look when you get the chance
